File: jvm/spring/nexus-core/nexus-core-aegis/src/main/java/org/nexus/core/aegis/kernel/ModelChecker.java
package org.nexus.core.aegis.kernel;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Deterministic state-space model checker — faithful port of
 * typescript/aegis-srv/src/model-checker.ts.
 *
 * <p>Runs a bounded, deterministic check over the structured state graph
 * (states with initial/terminal flags, transitions with from/to, invariants,
 * properties, temporal properties). Performs reachability analysis, deadlock
 * detection, unreachable-state discovery, and best-effort structural
 * invariant/property/temporal verdicts with counterexample traces. Free-form
 * guard/action/invariant text is treated structurally (identifier-reference
 * validation) — it is arbitrary TLA+ with no evaluator here.
 *
 * <p>Pure: no DB, no I/O. Unit-testable and used by both the structural
 * engine and (via the REST surface) the model-check endpoint.
 */
public final class ModelChecker {

    private ModelChecker() {}

    // ── Model shapes (JSON-shaped records, mirroring the TS interfaces) ──

    public record McState(String id, String name, boolean isInitial, boolean isTerminal) {}

    public record McTransition(String id, String name, String fromStateId, String toStateId,
                               String guardExpression, boolean weakFairness, boolean strongFairness,
                               Integer priority) {}

    public record McInvariant(String id, String name, String expression, boolean isTypeInvariant) {}

    public record McProperty(String id, String name, String type, String expression) {}

    public record McTemporalProperty(String id, String name, String operator, String expression) {}

    public record McModel(List<McState> states, List<McTransition> transitions,
                          List<McInvariant> invariants, List<McProperty> properties,
                          List<McTemporalProperty> temporalProperties,
                          List<String> variables, List<String> constants) {}

    // ── Verdicts ─────────────────────────────────────────────────────────

    public record Verdict(String kind, String name, String type, String result, String detail) {}

    public record Report(String status, List<String> reachableStates, List<String> unreachableStates,
                         List<String> deadlockTrace, List<Verdict> verdicts,
                         List<String> errors, List<String> warnings) {}

    // ── TLA+ identifier reference validation ─────────────────────────────

    private static final Set<String> TLA_RESERVED = Set.of(
            "MODULE", "CONSTANTS", "VARIABLES", "ASSUME", "THEOREM", "IN", "IF", "THEN",
            "ELSE", "CASE", "LET", "TRUE", "FALSE", "BOOLEAN", "CHOOSE", "EXCEPT",
            "DOMAIN", "UNION", "SUBSET", "LAMBDA", "Nat", "Int", "Real", "Seq", "Set",
            "NULL", "ENABLED", "UNCHANGED");

    private static final Pattern IDENTIFIER = Pattern.compile("[A-Za-z_][A-Za-z0-9_]*");
    private static final Pattern STRING_LITERAL = Pattern.compile("\"[^\"]*\"|'[^']*'");
    private static final Pattern BACKSLASH_OP = Pattern.compile("\\\\[A-Za-z]+");

    /** Extract bare identifiers from a TLA+ expression, skipping string literals and backslash-operators. */
    static List<String> identifiers(String expr) {
        if (expr == null || expr.isBlank()) return List.of();
        String cleaned = BACKSLASH_OP.matcher(STRING_LITERAL.matcher(expr).replaceAll(" ")).replaceAll(" ");
        List<String> out = new ArrayList<>();
        Matcher m = IDENTIFIER.matcher(cleaned);
        while (m.find()) {
            String s = m.group();
            if (!TLA_RESERVED.contains(s) && !"id".equals(s) && !"name".equals(s)
                    && !"type".equals(s) && !"status".equals(s)) {
                out.add(s);
            }
        }
        return out;
    }

    // ── Pure checker ─────────────────────────────────────────────────────

    public static Report checkModel(McModel model) {
        List<String> errors = new ArrayList<>();
        List<String> warnings = new ArrayList<>();
        List<Verdict> verdicts = new ArrayList<>();
        String[] status = {"success"};

        Map<String, McState> stateById = new LinkedHashMap<>();
        Map<String, String> stateNameById = new LinkedHashMap<>();
        for (McState s : model.states()) {
            stateById.put(s.id(), s);
            stateNameById.put(s.id(), s.name());
        }
        Set<String> knownNames = new LinkedHashSet<>();
        for (McState s : model.states()) knownNames.add(s.name());
        knownNames.addAll(model.variables());
        knownNames.addAll(model.constants());

        // Adjacency: stateId -> outgoing transitions. Wildcards (no from_state_id) apply everywhere.
        Map<String, List<McTransition>> outgoing = new LinkedHashMap<>();
        List<McTransition> wildcard = new ArrayList<>();
        for (McState s : model.states()) outgoing.put(s.id(), new ArrayList<>());
        for (McTransition t : model.transitions()) {
            if (t.fromStateId() != null && outgoing.containsKey(t.fromStateId())) {
                outgoing.get(t.fromStateId()).add(t);
            } else if (t.fromStateId() == null) {
                wildcard.add(t);
            }
        }

        // Structural validation
        List<McState> initial = model.states().stream().filter(McState::isInitial).toList();
        if (initial.isEmpty()) {
            status[0] = "failure";
            errors.add("no initial state declared (is_initial=false on all states)");
        }
        for (McTransition t : model.transitions()) {
            if (t.fromStateId() != null && !stateById.containsKey(t.fromStateId())) {
                errors.add("transition \"" + t.name() + "\" references missing from-state " + t.fromStateId());
            }
            if (t.toStateId() != null && !stateById.containsKey(t.toStateId())) {
                errors.add("transition \"" + t.name() + "\" references missing to-state " + t.toStateId());
            }
            if (t.fromStateId() != null && t.fromStateId().equals(t.toStateId())
                    && (t.guardExpression() == null || t.guardExpression().isBlank())) {
                warnings.add("transition \"" + t.name() + "\" is an unguarded self-loop (may not terminate)");
            }
        }

        // Reachability (BFS from all initial states)
        Set<String> reachable = new LinkedHashSet<>();
        Map<String, String> parent = new LinkedHashMap<>();
        List<String> queue = new ArrayList<>();
        for (McState s : initial) {
            reachable.add(s.id());
            parent.put(s.id(), null);
            queue.add(s.id());
        }
        for (int i = 0; i < queue.size(); i++) {
            String cur = queue.get(i);
            List<McTransition> edges = new ArrayList<>(outgoing.getOrDefault(cur, List.of()));
            edges.addAll(wildcard);
            for (McTransition t : edges) {
                String nextId = t.toStateId();
                if (nextId != null && stateById.containsKey(nextId) && !reachable.contains(nextId)) {
                    reachable.add(nextId);
                    parent.put(nextId, cur);
                    queue.add(nextId);
                }
            }
        }

        List<String> reachableNames = model.states().stream().filter(s -> reachable.contains(s.id()))
                .map(McState::name).toList();
        List<String> unreachableNames = model.states().stream().filter(s -> !reachable.contains(s.id()))
                .map(McState::name).toList();
        if (!unreachableNames.isEmpty()) {
            warnings.add("unreachable states: " + String.join(", ", unreachableNames));
        }

        // pathTo via BFS parents (state names): unshift the current node's name,
        // then walk to its parent until the chain ends at an initial state
        // (parent entry null). Yields [initial, ..., target].
        java.util.function.Function<String, List<String>> pathTo = targetId -> {
            List<String> path = new ArrayList<>();
            String cur = targetId;
            while (cur != null) {
                path.add(0, stateNameById.getOrDefault(cur, cur));
                cur = parent.get(cur);
            }
            return path;
        };

        // Deadlock detection: reachable non-terminal state with no outgoing edges
        String deadlock = null;
        for (String id : reachable) {
            McState s = stateById.get(id);
            boolean hasOut = !outgoing.getOrDefault(id, List.of()).isEmpty() || !wildcard.isEmpty();
            if (!hasOut && !s.isTerminal()) {
                deadlock = id;
                break;
            }
        }
        List<String> deadlockTrace = null;
        if (deadlock != null) {
            status[0] = "failure";
            deadlockTrace = pathTo.apply(deadlock);
            errors.add("deadlock: reachable non-terminal state \"" + stateNameById.get(deadlock)
                    + "\" has no outgoing transitions");
        }

        // Invariant checks (structural)
        for (McInvariant inv : model.invariants()) {
            if (inv.expression() == null || inv.expression().isBlank()) {
                verdicts.add(new Verdict("invariant", inv.name(), null, "FAIL", "empty invariant expression"));
                continue;
            }
            List<String> ids = identifiers(inv.expression());
            List<String> unknown = ids.stream().filter(s -> !knownNames.contains(s)).toList();
            if (!unknown.isEmpty()) {
                verdicts.add(new Verdict("invariant", inv.name(), null, "FAIL",
                        "references undefined identifier(s): " + String.join(", ", unknown)));
                if (!"failure".equals(status[0])) status[0] = "failure";
                continue;
            }
            if (inv.isTypeInvariant() && model.variables().stream().noneMatch(ids::contains)) {
                verdicts.add(new Verdict("invariant", inv.name(), null, "WARN",
                        "type invariant does not reference any declared variable (structural check cannot confirm)"));
                continue;
            }
            verdicts.add(new Verdict("invariant", inv.name(), null, "PASS",
                    "references only known identifiers (structural)"));
        }

        // Property checks (structural)
        for (McProperty prop : model.properties()) {
            List<String> referencedStates = identifiers(prop.expression()).stream()
                    .filter(knownNames::contains).toList();
            boolean referencedReachable = referencedStates.stream().allMatch(reachableNames::contains);
            switch (prop.type() == null ? "" : prop.type()) {
                case "safety" -> {
                    if (referencedStates.isEmpty()) {
                        verdicts.add(new Verdict("property", prop.name(), prop.type(), "PASS",
                                "safety property references no state (trivially structural)"));
                    } else if (referencedReachable) {
                        verdicts.add(new Verdict("property", prop.name(), prop.type(), "WARN",
                                "safety property references reachable state(s); needs TLA+ evaluation to prove"));
                    } else {
                        verdicts.add(new Verdict("property", prop.name(), prop.type(), "PASS",
                                "safety property only references unreachable state(s)"));
                    }
                }
                case "liveness" -> {
                    boolean reachableTerminal = model.states().stream()
                            .anyMatch(s -> reachable.contains(s.id()) && s.isTerminal());
                    boolean hasCycle = hasCycleInReachable(reachable, outgoing, wildcard);
                    if (referencedReachable && (reachableTerminal || hasCycle)) {
                        verdicts.add(new Verdict("property", prop.name(), prop.type(), "PASS",
                                "liveness: progress possible (reachable terminal or cycle present)"));
                    } else if (referencedReachable) {
                        verdicts.add(new Verdict("property", prop.name(), prop.type(), "FAIL",
                                "liveness: no reachable terminal and no cycle — progress cannot be satisfied"));
                        if (!"failure".equals(status[0])) status[0] = "failure";
                    } else {
                        verdicts.add(new Verdict("property", prop.name(), prop.type(), "WARN",
                                "liveness references unreachable state(s)"));
                    }
                }
                case "fairness" -> {
                    boolean hasFairness = model.transitions().stream()
                            .anyMatch(t -> t.weakFairness() || t.strongFairness());
                    verdicts.add(new Verdict("property", prop.name(), prop.type(),
                            hasFairness ? "PASS" : "WARN",
                            hasFairness ? "at least one transition declares weak/strong fairness"
                                    : "no transition declares fairness — cannot guarantee fairness"));
                }
                default -> verdicts.add(new Verdict("property", prop.name(), prop.type(), "WARN",
                        "unknown property type \"" + prop.type() + "\""));
            }
        }

        // Temporal property checks (structural)
        for (McTemporalProperty tp : model.temporalProperties()) {
            List<String> referencedStates = identifiers(tp.expression()).stream()
                    .filter(knownNames::contains).toList();
            boolean referencedReachable = referencedStates.stream().allMatch(reachableNames::contains);
            switch (tp.operator() == null ? "" : tp.operator()) {
                case "[]" -> verdicts.add(new Verdict("temporal", tp.name(), null,
                        referencedReachable ? "PASS" : "WARN",
                        referencedReachable ? "always: all referenced states are reachable (structural)"
                                : "always: references unreachable state(s)"));
                case "<>" -> {
                    if (referencedStates.isEmpty()) {
                        verdicts.add(new Verdict("temporal", tp.name(), null, "PASS",
                                "eventually: no referenced state (trivially structural)"));
                    } else {
                        verdicts.add(new Verdict("temporal", tp.name(), null,
                                referencedReachable ? "PASS" : "FAIL",
                                referencedReachable ? "eventually: referenced state is reachable"
                                        : "eventually: referenced state is NOT reachable"));
                        if (!referencedReachable && !"failure".equals(status[0])) status[0] = "failure";
                    }
                }
                case "->", "~>", "=>" -> verdicts.add(new Verdict("temporal", tp.name(), null,
                        referencedReachable ? "PASS" : "WARN",
                        referencedReachable ? "leads-to/implies: referenced states reachable (structural)"
                                : "leads-to/implies: references unreachable state(s)"));
                default -> verdicts.add(new Verdict("temporal", tp.name(), null, "WARN",
                        "unknown temporal operator \"" + tp.operator() + "\""));
            }
        }

        return new Report(status[0], reachableNames, unreachableNames, deadlockTrace, verdicts, errors, warnings);
    }

    /** Detect a cycle among reachable states (DFS over outgoing + wildcard edges). */
    private static boolean hasCycleInReachable(Set<String> reachable,
                                               Map<String, List<McTransition>> outgoing,
                                               List<McTransition> wildcard) {
        final int WHITE = 0, GRAY = 1, BLACK = 2;
        Map<String, Integer> color = new LinkedHashMap<>();
        for (String id : reachable) color.put(id, WHITE);

        for (String id : reachable) {
            if (color.get(id) != WHITE) continue;
            // Iterative DFS with explicit stack (node, edgeIndex).
            List<String> stack = new ArrayList<>();
            List<Integer> idx = new ArrayList<>();
            color.put(id, GRAY);
            stack.add(id);
            idx.add(0);
            while (!stack.isEmpty()) {
                int top = stack.size() - 1;
                String cur = stack.get(top);
                List<McTransition> edges = new ArrayList<>(outgoing.getOrDefault(cur, List.of()));
                edges.addAll(wildcard);
                if (idx.get(top) >= edges.size()) {
                    color.put(cur, BLACK);
                    stack.remove(top);
                    idx.remove(top);
                    continue;
                }
                McTransition t = edges.get(idx.get(top));
                idx.set(top, idx.get(top) + 1);
                String next = t.toStateId();
                if (next == null || !color.containsKey(next)) continue;
                int c = color.get(next);
                if (c == GRAY) return true;
                if (c == WHITE) {
                    color.put(next, GRAY);
                    stack.add(next);
                    idx.add(0);
                }
            }
        }
        return false;
    }
}

File: moleculer/aegis/services/model-checker.ts
/**
 * aegis-srv — deterministic state-space model checker.
 *
 * Runs a bounded, deterministic model check over the *structured* state graph
 * encoded in the `aegis` schema (states with is_initial/is_terminal,
 * transitions with from_state_id/to_state_id, invariants, properties,
 * temporal_properties). It performs genuine reachability analysis, deadlock
 * detection, unreachable-state discovery, and best-effort structural
 * invariant/property/temporal verdicts, producing counterexample traces.
 *
 * This is a real model-check baseline that does NOT require the TLA+ model
 * checker (TLC). A future TLC integration can augment it to evaluate the
 * free-form guard/action/invariant text; here those expressions are treated
 * structurally (identifier-reference validation) because they are arbitrary
 * TLA+ text with no evaluator in this environment.
 */

// ── Input model (plain JSON, DB-independent — unit-testable) ─────────

export interface MCState {
  id: string;
  name: string;
  is_initial: boolean;
  is_terminal: boolean;
  variable_assignments?: Record<string, unknown>;
}

export interface MCTransition {
  id: string;
  name: string;
  from_state_id: string | null;
  to_state_id: string | null;
  guard_expression?: string | null;
  action?: Record<string, unknown>;
  weak_fairness?: boolean;
  strong_fairness?: boolean;
  priority?: number;
}

export interface MCInvariant {
  id: string;
  name: string;
  expression: string;
  is_type_invariant: boolean;
}

export interface MCProperty {
  id: string;
  name: string;
  type: string; // safety | liveness | fairness
  expression: string;
}

export interface MCTemporalProperty {
  id: string;
  name: string;
  operator: string; // [] | <> | -> | ~> | =>
  expression: string;
}

export interface MCModel {
  states: MCState[];
  transitions: MCTransition[];
  invariants: MCInvariant[];
  properties: MCProperty[];
  temporal_properties: MCTemporalProperty[];
  variables?: string[];
  constants?: string[];
}

// ── Verdicts ─────────────────────────────────────────────────────────

export interface CheckVerdict {
  kind: 'invariant' | 'property' | 'temporal';
  name: string;
  type?: string;
  result: 'PASS' | 'FAIL' | 'WARN';
  detail: string;
}

export interface ModelCheckReport {
  /** success | failure */
  status: string;
  reachableStates: string[];
  unreachableStates: string[];
  /** counterexample path (state names) for a deadlock, if any */
  deadlockTrace?: string[];
  /** per-invariant / per-property / per-temporal verdicts */
  verdicts: CheckVerdict[];
  /** human-readable error list (empty on success) */
  errors: string[];
  warnings: string[];
}

// ── TLA+ identifier reference validation ─────────────────────────────

// TLA+ reserved words / common operators / standard module symbols. Any
// bare identifier in an expression not in this set must resolve to a known
// state name, variable name, or constant name.
const TLA_RESERVED = new Set([
  'MODULE', 'CONSTANTS', 'VARIABLES', 'ASSUME', 'THEOREM', 'IN', 'IF', 'THEN',
  'ELSE', 'CASE', 'LET', 'IN', 'TRUE', 'FALSE', 'BOOLEAN', 'CHOOSE', 'EXCEPT',
  'DOMAIN', 'UNION', 'SUBSET', 'LAMBDA', '\\in', '\\notin', '\\subseteq',
  '\\union', '\\intersect', '\\cup', '\\cap', '\\div', '\\o', '\\A', '\\E',
  '\\AA', '\\EE', '\\lnot', '\\land', '\\lor', '\\implies', '\\equiv',
  '\\x', '\\times', '\\geq', '\\leq', '\\neq', '=', '#', ':=', '->', '<->',
  '\\cdot', 'Nat', 'Int', 'Real', 'Seq', 'Set', 'NULL', 'ENABLED', 'UNCHANGED',
]);

const IDENTIFIER_RE = /[A-Za-z_][A-Za-z0-9_]*/g;

/** Extract bare identifiers from a TLA+ expression, ignoring string literals and backslash-operators. */
function identifiers(expr: string): string[] {
  // Strip single/double-quoted string literals and backslash-prefixed operators
  // (\in, \land, ...) so their contents / trailing tokens are not matched.
  const cleaned = expr
    .replace(/"[^"]*"/g, ' ')
    .replace(/'[^']*'/g, ' ')
    .replace(/\\[A-Za-z]+/g, ' ');
  return (cleaned.match(IDENTIFIER_RE) || []).filter(
    (s) => !TLA_RESERVED.has(s) && !['id', 'name', 'type', 'status'].includes(s),
  );
}

// ── Pure checker (no DB / no I/O) ─────────────────────────────────────

/**
 * Run the deterministic model check over a plain model. Pure and unit-testable.
 */
export function checkModel(model: MCModel): ModelCheckReport {
  const report: ModelCheckReport = {
    status: 'success',
    reachableStates: [],
    unreachableStates: [],
    verdicts: [],
    errors: [],
    warnings: [],
  };

  const stateById = new Map(model.states.map((s) => [s.id, s]));
  const stateNameById = new Map(model.states.map((s) => [s.id, s.name]));
  const knownNames = new Set([
    ...model.states.map((s) => s.name),
    ...(model.variables || []),
    ...(model.constants || []),
  ]);

  // Build adjacency: stateId -> outgoing transition ids. Wildcard transitions
  // (no from_state_id) apply to every state.
  const outgoing = new Map<string, MCTransition[]>();
  const wildcard = model.transitions.filter((t) => !t.from_state_id);
  for (const s of model.states) outgoing.set(s.id, []);
  for (const t of model.transitions) {
    if (t.from_state_id && outgoing.has(t.from_state_id)) {
      outgoing.get(t.from_state_id)!.push(t);
    }
  }

  // ── Structural validation ──────────────────────────────────────────
  const initial = model.states.filter((s) => s.is_initial);
  if (initial.length === 0) {
    report.status = 'failure';
    report.errors.push('no initial state declared (is_initial=false on all states)');
  }
  for (const t of model.transitions) {
    if (t.from_state_id && !stateById.has(t.from_state_id)) {
      report.errors.push(`transition "${t.name}" references missing from-state ${t.from_state_id}`);
    }
    if (t.to_state_id && !stateById.has(t.to_state_id)) {
      report.errors.push(`transition "${t.name}" references missing to-state ${t.to_state_id}`);
    }
    if (t.from_state_id && t.to_state_id && t.from_state_id === t.to_state_id && !t.guard_expression) {
      report.warnings.push(`transition "${t.name}" is an unguarded self-loop (may not terminate)`);
    }
  }

  // ── Reachability (BFS from all initial states) ─────────────────────
  const reachable = new Set<string>();
  const queue: string[] = initial.map((s) => s.id);
  const parent = new Map<string, string | null>();
  for (const id of initial) parent.set(id.id, null);
  for (const id of queue) reachable.add(id);

  while (queue.length > 0) {
    const cur = queue.shift()!;
    const edges = [...(outgoing.get(cur) || []), ...wildcard];
    for (const t of edges) {
      const nextId = t.to_state_id;
      // Only follow transitions to a state that exists in the model.
      if (nextId && stateById.has(nextId) && !reachable.has(nextId)) {
        reachable.add(nextId);
        parent.set(nextId, cur);
        queue.push(nextId);
      }
    }
  }

  report.reachableStates = model.states.filter((s) => reachable.has(s.id)).map((s) => s.name);
  report.unreachableStates = model.states.filter((s) => !reachable.has(s.id)).map((s) => s.name);
  if (report.unreachableStates.length > 0) {
    report.warnings.push(`unreachable states: ${report.unreachableStates.join(', ')}`);
  }

  const pathTo = (targetId: string): string[] => {
    const path: string[] = [];
    let cur: string | null = targetId;
    while (cur) {
      path.unshift(stateNameById.get(cur) || cur);
      cur = parent.get(cur) || null;
    }
    return path;
  };

  // ── Deadlock detection: reachable non-terminal state with no outgoing ─
  let deadlock: string | undefined;
  for (const id of reachable) {
    const s = stateById.get(id)!;
    const edges = outgoing.get(id) || [];
    const hasOut = edges.length > 0 || wildcard.length > 0;
    if (!hasOut && !s.is_terminal) {
      deadlock = id;
      break;
    }
  }
  if (deadlock) {
    report.status = 'failure';
    report.deadlockTrace = pathTo(deadlock);
    report.errors.push(`deadlock: reachable non-terminal state "${stateNameById.get(deadlock)}" has no outgoing transitions`);
  }

  // ── Invariant checks (structural) ──────────────────────────────────
  for (const inv of model.invariants) {
    if (!inv.expression || !inv.expression.trim()) {
      report.verdicts.push({ kind: 'invariant', name: inv.name, result: 'FAIL', detail: 'empty invariant expression' });
      continue;
    }
    const unknown = identifiers(inv.expression).filter((s) => !knownNames.has(s));
    if (unknown.length > 0) {
      report.verdicts.push({
        kind: 'invariant', name: inv.name, result: 'FAIL',
        detail: `references undefined identifier(s): ${unknown.join(', ')}`,
      });
      if (report.status !== 'failure') report.status = 'failure';
      continue;
    }
    // type invariants must reference a variable or constant by name
    if (inv.is_type_invariant && !(model.variables || []).some((v) => identifiers(inv.expression).includes(v))) {
      report.verdicts.push({
        kind: 'invariant', name: inv.name, result: 'WARN',
        detail: 'type invariant does not reference any declared variable (structural check cannot confirm)',
      });
      continue;
    }
    report.verdicts.push({ kind: 'invariant', name: inv.name, result: 'PASS', detail: 'references only known identifiers (structural)' });
  }

  // ── Property checks (structural) ───────────────────────────────────
  for (const prop of model.properties) {
    const referencedStates = identifiers(prop.expression).filter((s) => knownNames.has(s));
    const referencedReachable = referencedStates.every((s) => report.reachableStates.includes(s));
    if (prop.type === 'safety') {
      // A safety property "never in state X" holds if X is unreachable; if X
      // is reachable we cannot prove it without an evaluator -> WARN.
      if (referencedStates.length === 0) {
        report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'PASS', detail: 'safety property references no state (trivially structural)' });
      } else if (referencedReachable) {
        report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'WARN', detail: 'safety property references reachable state(s); needs TLA+ evaluation to prove' });
      } else {
        report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'PASS', detail: 'safety property only references unreachable state(s)' });
      }
    } else if (prop.type === 'liveness') {
      // Liveness "eventually ..." needs progress: a reachable terminal OR a cycle.
      const reachableTerminal = model.states.some((s) => reachable.has(s.id) && s.is_terminal);
      const hasCycle = hasCycleInReachable(reachable, outgoing, wildcard, stateById);
      if (referencedReachable && (reachableTerminal || hasCycle)) {
        report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'PASS', detail: 'liveness: progress possible (reachable terminal or cycle present)' });
      } else if (referencedReachable) {
        report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'FAIL', detail: 'liveness: no reachable terminal and no cycle — progress cannot be satisfied' });
        if (report.status !== 'failure') report.status = 'failure';
      } else {
        report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'WARN', detail: 'liveness references unreachable state(s)' });
      }
    } else if (prop.type === 'fairness') {
      const hasFairness = model.transitions.some((t) => t.weak_fairness || t.strong_fairness);
      report.verdicts.push({
        kind: 'property', name: prop.name, type: prop.type,
        result: hasFairness ? 'PASS' : 'WARN',
        detail: hasFairness ? 'at least one transition declares weak/strong fairness' : 'no transition declares fairness — cannot guarantee fairness',
      });
    } else {
      report.verdicts.push({ kind: 'property', name: prop.name, type: prop.type, result: 'WARN', detail: `unknown property type "${prop.type}"` });
    }
  }

  // ── Temporal property checks (structural) ──────────────────────────
  for (const tp of model.temporal_properties) {
    const referencedStates = identifiers(tp.expression).filter((s) => knownNames.has(s));
    const referencedReachable = referencedStates.every((s) => report.reachableStates.includes(s));
    switch (tp.operator) {
      case '[]': // always
        report.verdicts.push({
          kind: 'temporal', name: tp.name, result: referencedReachable ? 'PASS' : 'WARN',
          detail: referencedReachable ? 'always: all referenced states are reachable (structural)' : 'always: references unreachable state(s)',
        });
        break;
      case '<>': // eventually
        if (referencedStates.length === 0) {
          report.verdicts.push({ kind: 'temporal', name: tp.name, result: 'PASS', detail: 'eventually: no referenced state (trivially structural)' });
        } else {
          report.verdicts.push({
            kind: 'temporal', name: tp.name,
            result: referencedReachable ? 'PASS' : 'FAIL',
            detail: referencedReachable ? 'eventually: referenced state is reachable' : 'eventually: referenced state is NOT reachable',
          });
          if (!referencedReachable && report.status !== 'failure') report.status = 'failure';
        }
        break;
      case '->':
      case '~>': // leads-to
      case '=>': // implies
        report.verdicts.push({
          kind: 'temporal', name: tp.name,
          result: referencedReachable ? 'PASS' : 'WARN',
          detail: referencedReachable ? 'leads-to/implies: referenced states reachable (structural)' : 'leads-to/implies: references unreachable state(s)',
        });
        break;
      default:
        report.verdicts.push({ kind: 'temporal', name: tp.name, result: 'WARN', detail: `unknown temporal operator "${tp.operator}"` });
    }
  }

  return report;
}

/** Detect a cycle among reachable states (DFS over outgoing + wildcard edges). */
function hasCycleInReachable(
  reachable: Set<string>,
  outgoing: Map<string, MCTransition[]>,
  wildcard: MCTransition[],
  stateById: Map<string, MCState>,
): boolean {
  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Map<string, number>();
  for (const id of reachable) color.set(id, WHITE);

  const visit = (id: string): boolean => {
    color.set(id, GRAY);
    const edges = [...(outgoing.get(id) || []), ...wildcard];
    for (const t of edges) {
      if (!t.to_state_id) continue;
      if (!color.has(t.to_state_id)) continue; // not reachable -> skip
      const c = color.get(t.to_state_id)!;
      if (c === GRAY) return true; // back edge -> cycle
      if (c === WHITE && visit(t.to_state_id)) return true;
    }
    color.set(id, BLACK);
    return false;
  };

  for (const id of reachable) {
    if (color.get(id) === WHITE && visit(id)) return true;
  }
  return false;
}